File: usedesk_sdk/src/main/java/ru/usedesk/sdk/external/ui/knowledgebase/pages/sections/SectionsAdapter.java
package ru.usedesk.sdk.external.ui.knowledgebase.pages.sections;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;

import java.util.List;

import ru.usedesk.sdk.R;
import ru.usedesk.sdk.external.entity.knowledgebase.Section;
import ru.usedesk.sdk.external.ui.UsedeskViewCustomizer;

public class SectionsAdapter extends RecyclerView.Adapter<SectionsAdapter.SectionViewHolder> {

    private final List<Section> sectionList;
    private final IOnSectionClickListener onSectionClickListener;
    private final UsedeskViewCustomizer usedeskViewCustomizer;

    SectionsAdapter(@NonNull List<Section> sectionList,
                    @NonNull IOnSectionClickListener onSectionClickListener,
                    @NonNull UsedeskViewCustomizer usedeskViewCustomizer) {
        this.sectionList = sectionList;
        this.onSectionClickListener = onSectionClickListener;
        this.usedeskViewCustomizer = usedeskViewCustomizer;
    }

    @NonNull
    @Override
    public SectionViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View view = usedeskViewCustomizer.createView(viewGroup, R.layout.usedesk_item_section);

        return new SectionViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull SectionViewHolder sectionViewHolder, int i) {
        sectionViewHolder.bind(sectionList.get(i));
    }

    @Override
    public int getItemCount() {
        return sectionList.size();
    }

    class SectionViewHolder extends RecyclerView.ViewHolder {

        private final View rootView;
        private final ImageView imageViewIcon;
        private final TextView textViewTitle;

        SectionViewHolder(@NonNull View itemView) {
            super(itemView);

            rootView = itemView;
            imageViewIcon = itemView.findViewById(R.id.iv_icon);
            textViewTitle = itemView.findViewById(R.id.tv_title);
        }

        void bind(@NonNull final Section section) {
            imageViewIcon.setImageBitmap(null);
            Glide.with(imageViewIcon)
                    .load(section.getImage())
                    .into(imageViewIcon);
            textViewTitle.setText(section.getTitle());

            rootView.setOnClickListener(v -> onSectionClickListener.onSectionClick(section.getId()));
        }
    }
}

File: knowledgebase-gui/src/main/java/ru/usedesk/knowledgebase_gui/pages/sections/SectionsAdapter.kt
package ru.usedesk.knowledgebase_gui.pages.sections

import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.lifecycle.LifecycleOwner
import androidx.recyclerview.widget.RecyclerView
import ru.usedesk.common_gui.UsedeskBinding
import ru.usedesk.common_gui.inflateItem
import ru.usedesk.common_gui.showImage
import ru.usedesk.knowledgebase_gui.R
import ru.usedesk.knowledgebase_sdk.entity.UsedeskSection

internal class SectionsAdapter(
        recyclerView: RecyclerView,
        lifecycleOwner: LifecycleOwner,
        private val viewModel: SectionsViewModel,
        private val onSectionClick: (Long) -> Unit
) : RecyclerView.Adapter<SectionsAdapter.SectionViewHolder>() {

    private var sectionList = listOf<UsedeskSection>()

    init {
        recyclerView.adapter = this

        viewModel.sectionsLiveData.observe(lifecycleOwner) {
            this.sectionList = it
            notifyDataSetChanged()
        }
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): SectionViewHolder {
        val binding = inflateItem(viewGroup,
                R.layout.usedesk_item_section,
                R.style.Usedesk_KnowledgeBase) { rootView, defaultStyleId ->
            Binding(rootView, defaultStyleId)
        }
        return SectionViewHolder(binding)
    }

    override fun onBindViewHolder(sectionViewHolder: SectionViewHolder, i: Int) {
        sectionViewHolder.bind(sectionList[i])
    }

    override fun getItemCount(): Int = sectionList.size

    inner class SectionViewHolder(
            private val binding: Binding
    ) : RecyclerView.ViewHolder(binding.rootView) {

        fun bind(section: UsedeskSection) {
            binding.ivIcon.setImageBitmap(null)
            binding.tvTitle.text = section.title
            binding.ivIcon.setImageResource(R.drawable.background_no_thumbnail)
            binding.tvInitials.text = section.title
            section.thumbnail?.also {
                showImage(binding.ivIcon,
                        R.drawable.background_no_thumbnail,
                        it,
                        onSuccess = {
                            binding.tvInitials.text = ""
                        }
                )
            }
            binding.rootView.setOnClickListener {
                onSectionClick(section.id)
            }
        }
    }

    internal class Binding(rootView: View, defaultStyleId: Int) : UsedeskBinding(rootView, defaultStyleId) {
        val ivIcon: ImageView = rootView.findViewById(R.id.iv_icon)
        val tvTitle: TextView = rootView.findViewById(R.id.tv_title)
        val tvInitials: TextView = rootView.findViewById(R.id.tv_initials)
    }
}